# Learning From Data - Homework 6
## Ognen Nastov

![](hw6_images/hw6p1.png)

**Answer:**

When target is fixed, and $H$ complexity ⬆️, deterministic noise ⬇️.

$H'$ is less complex than $H$.

=> Deterministic noise will increase.

Answer is [b].

---

![](hw6_images/hw6p2.png)

**Answer:**

In [1]:
import numpy as np

# problem 2 
# read input and test sets
# d = 2
# return X,Y where X = [[x1, ...] and Y = [y1, ...]
def read_training_set():
    S = np.loadtxt("http://work.caltech.edu/data/in.dta")
    return S[:,0:2], S[:,2]

def read_test_set():
    S = np.loadtxt("http://work.caltech.edu/data/out.dta")
    return S[:,0:2], S[:,2]

# transformed input set is d = 7
def transform_input_set(X):
    N = np.size(X, 0)
    X0 = np.ones([N,])
    X1 = X[:,0]
    X2 = X[:,1]
    X3 = X1**2
    X4 = X2**2
    X5 = X1*X2
    X6 = np.abs(X1-X2)
    X7 = np.abs(X1+X2)
    X_transformed = np.array([X0, X1, X2, X3, X4, X5, X6, X7]).T
    return (X_transformed)

In [2]:
def solve_regression(X, y):
    w = np.dot(np.linalg.pinv(X), y)
    return (w)

# classification error
def classification_error(w, X, y):
    y_c = np.sign(np.dot(X,w))
    error = np.count_nonzero(y_c != y) / len(y)
    return (error)

def problem_2():
    X,y = read_training_set()
    X_t,y_t = read_test_set()
    X_transformed = transform_input_set(X)
    X_t_transformed = transform_input_set(X_t)
    w = solve_regression(X_transformed, y)
    # E_in (classification error)
    E_in = classification_error(w, X_transformed, y)
    # E_out (test set classification error)
    E_out = classification_error(w, X_t_transformed, y_t)
    print(f"E_in = {E_in} , E_out = {E_out}")
    return

In [3]:
problem_2()

E_in = 0.02857142857142857 , E_out = 0.084


Answer is [a].

---

![](hw6_images/hw6p3a.png)
![](hw6_images/hw6p3b.png)

**Answer:**

In [4]:
# problems 3, 4, 5, 6
def solve_regression_with_weight_decay(X, y, k):
    _lambda = np.power(10.0, k)
    Q = np.size(X, 1)
    X_pinv = np.linalg.inv(X.T@X + _lambda*np.eye(Q))@X.T
    w = np.dot(X_pinv, y)
    return w

def problems_3_to_6(k):
    X,y = read_training_set()
    X_t,y_t = read_test_set()
    X_transformed = transform_input_set(X)
    X_t_transformed = transform_input_set(X_t)
    w = solve_regression_with_weight_decay(X_transformed, y, k)
    # E_in (classification error)
    E_in = classification_error(w, X_transformed, y)
    # E_out (test set classification error)
    E_out = classification_error(w, X_t_transformed, y_t)
    print(f"E_in = {E_in} , E_out = {E_out}")
    return

In [5]:
problems_3_to_6(-3)

E_in = 0.02857142857142857 , E_out = 0.08


Answer is [d].

---

![](hw6_images/hw6p4.png)

**Answer:**

In [6]:
problems_3_to_6(3)

E_in = 0.37142857142857144 , E_out = 0.436


Answer is [e].

---

![](hw6_images/hw6p5.png)

**Answer:**

In [7]:
problems_3_to_6(-2)

E_in = 0.02857142857142857 , E_out = 0.084


In [8]:
problems_3_to_6(-1)

E_in = 0.02857142857142857 , E_out = 0.056


In [9]:
problems_3_to_6(0)

E_in = 0.0 , E_out = 0.092


In [10]:
problems_3_to_6(1)

E_in = 0.05714285714285714 , E_out = 0.124


In [11]:
problems_3_to_6(2)

E_in = 0.2 , E_out = 0.228


Smallest $E_{out}$ for $k = -1$. Answer is [d].

---

![](hw6_images/hw6p6.png)

**Answer:**

Minimum for $k = -1$, $E_{out} = 0.056$

Answer is [b].

---

![](hw6_images/hw6p7a.png)
![](hw6_images/hw6p7b.png)

**Answer:**

If $C = 1$, $w_q = 1$ for $q ≥ Q_0$. But we need $w_q = 0$ for $q > Q$.

In [b], the second term has $Q_0 = 4$, meaning $w_4 = 1$. $Q = 3$, meaning we need $w_4 = 0$. The statement is incorrect.
 
In [d], the second term has $Q_0 = 4$, meaning $w_4 = 1$. $Q = 1$, meaning we need $w_2 = w_3 = w_4 = 0$. The statement is incorrect.

In [a], the first term has $Q_0 = 3$, meaning $w_3 = 0$. $Q = 4$, meaning we need $w_3$ to be nonzero. The statement is incorrect.

In [c], the first term has $Q_0 = 3$, meaning $w_3 = 0$, and the second term has $Q_0 = 4$, meaning $w_4 = 0$. $Q = 1$, meaning only $w_0$ and $w_1$ are nonzero. This statement is correct.

Answer is [c].

---

![](hw6_images/hw6p8.png)

**Answer:**

One iteration of backpropagation includes forward propagation, backpropagation, and the SGD update.

- Input layer `L0`: $d^{(0)}=5$ plus $1$ for the constant term = 6 nodes.
- `L1`: $d^{(1)}=3$ plus $1$ for constant term = 4 nodes.
- Output layer `L2`: $d^{(2)}=1$ = 1 node.

Draw the NN.

![](hw6_images/hw6_problem_8_sketch.png)

Going from `L0` to `L1`, there are 6\*3 = 18 links (since no links to constant term in `L1`).  
Going from `L1` to `L2`, there are 4 links.  
=> Forward propagation has 18+4=22 links, i.e. 22 weights.

SGD update needs to update all 22 weights.

Backpropagation. We don't need to:

- compute delta for the output (because it doesn't need any of the operations, counted as operation in this certain task).

- compute deltas for constants (because they're constants).

- compute deltas for input (because they're just features (x), and delta is a derivative of the error with respect to s).

So for computing deltas you only need 3 operations going from `L2` to `L1`.

=> Number of operations = 22 + 22 + 3 = 47.

Answer is 45 i.e. [d].

---

![](hw6_images/hw6p9a.png)
![](hw6_images/hw6p9b.png)

**Answer:**
    
We should have as many hidden layers as possible. Minimum number of units per layer is 2 (including constant node), so 36/2 = 18 hidden layers.

Number of weights (links) = 10\*(2-1) + 2\*(2-1) + ... + 2\*(2-1) = 10+18\*2 = 46

Answer is [a].

---

![](hw6_images/hw6p10.png)

**Answer:**

Assume two hidden layers, with $m$ and $n$ nodes. $i = 10$ (input nodes) and $m+n = s = 36$ total nodes.

It is intuitive to have 2 layers vs more when maximizing number of weights:  
- First, adding more layers creates more wasted constant nodes.
- Second, even if there were no wasted nodes we would be creating smaller multipliers and then summing them up vs multiplying them directly. i.e. adding layers cannot increase the weights among the hidden layers (it only deprives some nodes from connecting to others by putting them more than one layer apart) and cannot increase the connections to the input and output nodes.

Maximize number of weights (links): 

$$f(m,n) = i*(m-1) + m*(n-1) + n*1$$

Substitute $m = s-n$:

$$\begin{aligned}
f(m) & = i*(s-n-1) + (s-n)*(n-1) + n \\
& = -n^2 + (-i + s + 1 + 1)*n + i*s - i - s \\
& = -n^2 + (s + 2 - i)*n + (i*s - i - s)
\end{aligned}$$

Set $\frac{df}{dm} = 0$ to find extrema:

$$\frac{df}{dm} =  -2*n + s + 2 - i = 0$$

where from:

$$n = \frac{1}{2}(s+2-i)$$

Substituting $s=26$, $i=10$ gives $n = 14$ and $m = 22$.

Number of weights (links):

$$f(m,n) = 10*(22-1) + 22*(14-1) + 14 = 10*21+22*13+14 = 510$$

Answer is [e].

---